### ISSUE TITLE
Typescript compilation errors trigger infinite reloads

### ISSUE THREAD

**Issue by messman** [react-cosmos/react-cosmos#1353]
### What's wrong?

If strict Typescript checks are on while editing code, any TypeScript compilation errors will cause a fixture to reload indefinitely until compilation errors are fixed or until the browser tab is closed. The infinite reloading dramatically increases CPU usage by the browser and may cause crashes.

### Steps to reproduce

[See this test repo](https://github.com/messman/cosmos-reload-example). Instructions, as well as a discussion of the problem and research, are in the README.

Essentially:
- Set up a project for React Cosmos that uses strict TypeScript type checking.
- While viewing React Cosmos fixtures in a browser, cause a TypeScript compilation error in a file and save, triggering HMR.
- Watch as the test fixture starts reloading until stopped.

### User info

See the test repo for specifics on versions.
- React Cosmos 5.6.4
- Webpack 5.58.2
- TS Loader 9.2.6
- TypeScript 4.4.4

### Additional context

As discussed previously in #1315, I use strict type checking while working in React Cosmos. After working through performance issues, this is the only sore spot. The test repo linked above includes my research in tracking down the cause of the reloads to `@skidding/webpack-hot-middleware` ([webpack-contrib/webpack-hot-middleware](https://github.com/webpack-contrib/webpack-hot-middleware)), specifically the `reload` option that is hardcoded to `true`. This causes React Cosmos to operate more like `dev-server` than `only-dev-server` for HMR, reloading on errors and when there are no module updates.

I'll often fire up React Cosmos and then edit files to see live changes. Along the way, I'll commonly reorganize multiple files and end up creating many failed webpack builds from TypeScript compilation errors. After five minutes, I'll realize there's an issue with React Cosmos because my MacBook's fan will start getting louder and louder. I'll typically need to close the browser tab in order to calm the CPU down. 

In `plugins/webpack/webpackConfig/devServer.js`:

```js
function getHotMiddlewareEntry() {
  var clientPath = require.resolve('@skidding/webpack-hot-middleware/client');

  return "".concat(clientPath, "?reload=true&overlay=false");
}
```

Changing `reload=true` to `reload=false` or removing it entirely will stop the frame from reloading when a TypeScript compilation error occurs.

As the reloads occur, this HMR logging is printed in the browser console:

<img width="575" alt="agm 2021-10-17 at 1 28 39 AM" src="https://user-images.githubusercontent.com/5817216/137613187-24cf93ed-f415-476c-933a-8ad834681346.png">

```
[HMR] connected
[HMR] Checking for updates on the server...
Failed to load resource: the server responded with a status of 404 (Not Found)
[HMR] Cannot find update (Full reload needed)
[HMR] (Probably because of restarting the server)
[HMR] Reloading page
```

I hope my research above helps anyone else who has been stuck on this. It's not a show-stopper for me, but it's quite annoying. The `reload=false` workaround works well for me right now.

For a full solution, I would suggest the addition of an option in `cosmos.config.json` for controlling the `reload` behavior of HMR. I'd like to know what others think. Cheers.

**Comment by oliverschoning** [comment]
I have been able to avoid this problem in my TypeScript codebase by setting `ts-loader` to  `transpileOnly: true` and using `fork-ts-checker-webpack-plugin` with the `{ async: true }` option.

**Comment by messman** [comment]
This is now fixed and published in `react-cosmos@5.6.5`.

There is a new Cosmos config option under the `webpack` property named `reloadOnFail` defaulting to `true`. Specifying `false` will turn off location reloading on error and stop the infinite reloading issue.
The property will be available in the schema via SchemaStore/schemastore#1966.

```json
"reloadOnFail": {
  "description": "When using webpack's Hot Module Replacement, enable automatic location reload after webpack fails to update modules. [default: true]",
  "type": "boolean"
}
```

For more information on the fix, see #1358.

Thanks again for the help, @skidding!
